### PR TITLE
Enable specifying webkit-test-runner header values from a command-line option

### PR DESCRIPTION
**Pull request by sideshowbarker**
#### b3cdd4ed520fc96cff30af68e3ccc444a4c94a56
<pre>
Enable specifying webkit-test-runner header values from a command-line option
<a href="https://bugs.webkit.org/show_bug.cgi?id=264948">https://bugs.webkit.org/show_bug.cgi?id=264948</a>

Reviewed by Jonathan Bedard and Tim Horton.

This change adds a new “--additional-header” option to run-webkit-tests
(and WebKitTestRunner) to enable specifying webkit-test-runner Foo=bar
header values for test runs — as if those values were provided in a
“&lt;!-- webkit-test-runner [ Foo=Bar ] --&gt;” header in each test being run.

For example, given the following command-line invocation:

  run-webkit-tests --additional-header=&apos;useEphemeralSession=true&apos; PATH

…that invocation would cause the test(s) at the given PATH value to be
run (once) in an ephemeral session — that is, as if the test itself had a
“&lt;!-- webkit-test-runner [ useEphemeralSession=true ] --&gt;” header.

This differs from the “--self-compare-with-header” option, which causes each
test to be run twice — once with the given header value, and once without —
and then does a pixel-diff comparison between the results. That’s useful
for reftests, but not so useful for JavaScript tests or other test types.

In contrast to that, the “--additional-header” option causes the tests to be
run just once, with the given header value, and does no separate comparison
against the results without the given header value (other than just the
normal reporting whether the actual results matched the expected results).
So the option is generally useful for all test types, not just for reftests.

* Tools/Scripts/webkitpy/layout_tests/controllers/single_test_runner.py:
(SingleTestRunner._driver_input):
* Tools/Scripts/webkitpy/layout_tests/run_webkit_tests.py:
(parse_args):
* Tools/Scripts/webkitpy/layout_tests/views/printing.py:
(Printer.print_config):
* Tools/Scripts/webkitpy/layout_tests/views/printing_unittest.py:
(Testprinter.test_print_config):
* Tools/Scripts/webkitpy/port/driver.py:
(DriverInput.__init__):
(DriverInput.__repr__):
(Driver._command_from_driver_input):
* Tools/TestRunnerShared/TestCommand.cpp:
(WTR::parseInputLine):
* Tools/TestRunnerShared/TestCommand.h:
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::parseTestHeaderString):
(WTR::featureFromAdditionalHeaderOption):
* Tools/TestRunnerShared/TestFeatures.h:
* Tools/WebKitTestRunner/Options.cpp:
(WTR::handleOptionAdditionalHeader):
(WTR::OptionsHandler::OptionsHandler):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::testOptionsForTest const):

Canonical link: <a href="https://commits.webkit.org/273763@main">https://commits.webkit.org/273763@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e2ca5254577f8331ca5dff2b0af822f061f53a01

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26836 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28072 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29047 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24526 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27282 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2843 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24413 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27098 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4267 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23033 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3744 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26996 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3788 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29679 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24472 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30025 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3817 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2016 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27932 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5275 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8295 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4284 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4185 "Built successfully") | | |

### CODE REVIEW

**Comment by gsnedders** [comment]
And to note based on the Slack discussion that led to this, this differs to `--self-compare-with-header` which effectively runs each test twice and does a pixel diff between the two (which arguably isn't overly useful for JS tests!) because this simply runs the tests as normal, but with an extra header.

**Review by hortont424 — APPROVED** [review]

File: Tools/Scripts/webkitpy/port/driver.py
Comment (gsnedders) [inline Tools/Scripts/webkitpy/port/driver.py]: Given this is constructing a list to pass to `subprocess.Popen` or similar, don't we actually want:

```suggestion
            cmd.append("--additional-header")
            cmd.append(self._port.get_option('additional_header'))
```

(compare with the lines just below)

**Review by gsnedders — DISMISSED** [review]

**Comment by sideshowbarker** [comment]
> And to note based on the Slack discussion that led to this, this differs to `--self-compare-with-header` which effectively runs each test twice and does a pixel diff between the two (which arguably isn't overly useful for JS tests!) because this simply runs the tests as normal, but with an extra header.

Thanks — I updated the commit message (and the PR description) to add something along those lines.

File: Tools/Scripts/webkitpy/port/driver.py
Comment (sideshowbarker) [reply Tools/Scripts/webkitpy/port/driver.py]: > Given this is constructing a list to pass to `subprocess.Popen` or similar, don't we actually want:

After further testing, I see that this change is actually causing WebKitTestRunner to crash in the case where the
`--additional-header` argument has multiple values; for example: `--additional-header='foo=bar hoge=moge'` or whatever. So I’ll need to fix that before this can be merge-worthy

Comment (gsnedders) [reply Tools/Scripts/webkitpy/port/driver.py]: > So I’ll need to fix that before this can be merge-worthy

Given this is now resolved, is that fixed?

Comment (sideshowbarker) [reply Tools/Scripts/webkitpy/port/driver.py]: > > So I’ll need to fix that before this can be merge-worthy
> 
> Given this is now resolved, is that fixed?

No — it’s broken now. But it worked before without crashing, in the format-string form — which is surprising, if it’s getting passed to `subprocess.Popen` or similar.

But doing it instead the right way (or what would seem to be the right way) — that is, with multiple append()s — causes a crash (in the server, it seems) when `self._server_process.write(command)` is called.

And it crashes without producing any crash log — no stack trace, no error messages about where it actually fails.

I notice the thing that the existing code is already doing here:

https://github.com/WebKit/WebKit/blob/cb302cd1f44f09c8b330f4c25efb99347b3b8878/Tools/Scripts/webkitpy/port/driver.py#L666-L674

That is, `command` is a string, and that code is constructing it from the arguments using string concatenation, with the arguments separated from each other by `'` apostrophe characters.

And so I guess somewhere in the code it gets by the `self._server_process.write(command)` call, there is some code that parses out that string by splitting it on the `'` apostrophe characters — and it seems like that must be the code where the crash is happening. Not sure, because I haven’t looked there yet — but that’s where I’ll look next.

Anyway, I am guessing that maybe that code doesn’t expect to find a space character in the arguments that get split out from that string?

But so far from debugging, it looks to me like the string it ends up passing in the multiple-append()s case (committed from your suggestion) is actually identical to what it ends up passing from my original format-string form. At least from eyeballing a dump of the strings produced in each case, I can’t see any difference.

Anyway, I remain a bit baffled by it at this point, but as I mentioned, I will next look at the `_server_process.write(command)` code where the `command` string gets handled, and see if that ends up revealing anything more.

Comment (sideshowbarker) [reply Tools/Scripts/webkitpy/port/driver.py]: OK, I’ve fixed the crash. It turns out it’s not actually necessary to append `--additional-header` to the `cmd` value in that `cmd_line` function — the `--additional-header` option has the expected effect without even doing that.

As far as I could see, the only observable effect that appending `--additional-header` there has on anything is, to cause `--additional-header` and its value to be appended to the `Command line:` info message that gets emitted by the `print_config` function in the `Printer` logger.

So I instead just directly added some additional info logging to that `print_config` function — to explicitly echo out the `--additional-header` value that’s set in the `self._options.additional_header` value.

And if that all now looks OK, then I think this is otherwise merge-worthy.

**Comment by sideshowbarker** [comment]
@gsnedders After you’ve taken a look at this again, please let me know if you reckon there’s anything remaining that I need to do here

**Comment by sideshowbarker** [comment]
@gsnedders GitHub seems to still have this flagged as Changes Requested, even though https://github.com/WebKit/WebKit/pull/20608#pullrequestreview-1735290285 is resolved. Maybe I've missed something?

**Review by gsnedders — COMMENTED** [review]
[non-reviewer] this seems fine to me, now

**Comment by sideshowbarker** [comment]
@JonWBedard I believe this one is just awaiting code-owner review

**Review by JonWBedard — APPROVED** [review]

**Pull request merged**
